joblib/parallel.py

```diff
@@ -954,58 +1121,202 @@ def print_progress(self):
+                        time.sleep(0.01)
+                        continue
+
+                    # We need to be careful: the job list can be filling up as
+                    # we empty it and Python list are not thread-safe by
+                    # default hence the use of the lock
+                    with self._lock:
+                        batched_results = self._jobs.pop(0)
+
+                    # Flatten the batched results to output one output
+                    # at a time
+                    batched_results = batched_results.get_result(self.timeout)
+                    for result in batched_results:
+                        nb_consumed += 1
+                        yield result
+
+        # Note: we catch any BaseException instead of just
+        # Exception instances to also include KeyboardInterrupt
+        # and GeneratorExit
+        except BaseException as e:
+            self._exception = True
+            self._abort()
+
+            if self.return_generator and isinstance(e, GeneratorExit):
+                self._warn_exit_early(nb_consumed)
+            raise
+        finally:
+            _remaining_outputs = ([] if self._exception else self._jobs)
+            self._jobs = list()
+            self._running = False
+            self._terminate_and_reset()
+
+        while len(_remaining_outputs) > 0:
+            batched_results = _remaining_outputs.pop(0)
+            batched_results = batched_results.get_result(self.timeout)
+            for result in batched_results:
+                yield result
+
+    def _raise_error_fast(self):
+        """If we are aborting, raise if a job caused an error."""
+
+        # Find the first job whose status is TASK_ERROR if it exists.
+        with self._lock:
+            error_job = next((job for job in self._jobs
+                              if job.status == TASK_ERROR), None)
+
+        # If this error job exists, immediatly raise the error by
+        #  calling get_result. This jo might not exists if abort has been
+        # called directly or if the generator is gc.
+        if error_job is not None:
+            error_job.get_result(self.timeout)
+
+    def _warn_exit_early(self, nb_consumed):
+        ready_outputs = self.n_completed_tasks - nb_consumed
+        is_completed = self._is_completed()
+        msg = ""
+        if ready_outputs:
+            msg += ("%d tasks have been successfully executed and "
+                    "readied but not used." % ready_outputs)
+            if not is_completed:
+                msg += " Additionally, "
+
+        if not is_completed:
+            msg += ("%d tasks which were still being processed by the "
+                    "workers have been cancelled."
+                    % self.n_dispatched_tasks)
+            # Abort computation to avoid waiting unnecessarily for
+            # the results when they cannot be recovered.
+            self._abort()
```

Comment:
unnecessary, because the only call of `_warn_exit_early` is preceded by a call to `abort` ? https://github.com/joblib/joblib/blob/1062316d210f39f7fca44d5d29647267c91ee14a/joblib/parallel.py#L1208-L1211